File: swift/FPFundamentals/FPFundamentals.playground/Pages/Fundamentals.xcplaygroundpage/Contents.swift
import Bow

// Data types
class ForMaybe {}
typealias MaybeOf<A> = Kind<ForMaybe, A>

class Maybe<A>: MaybeOf<A> {
    static func fix(_ value: MaybeOf<A>) -> Maybe<A> {
        return value as! Maybe<A>
    }
    
    static func yes(_ a: A) -> Maybe<A> {
        return Yes(a)
    }
    
    static func no() -> Maybe<A> {
        return No()
    }
    
    func fold<B>(_ ifAbsent: () -> B, _ ifPresent: (A) -> B) -> B {
        switch self {
        case is No<A>: return ifAbsent()
        case is Yes<A>:
            let yes = self as! Yes
            return ifPresent(yes.value)
        default:
            fatalError("Unreachable code")
        }
    }
}

class Yes<A>: Maybe<A> {
    let value: A
    
    init(_ value: A) {
        self.value = value
    }
}

class No<A>: Maybe<A> {}

extension Kind where F == ForMaybe {
    func fix() -> Maybe<A> {
        return Maybe<A>.fix(self)
    }
}

extension Maybe: CustomStringConvertible {
    var description: String {
        return self.fold({ "No" },
                         { value in "Yes(\(value))" })
    }
}

// Typeclasses

protocol Combinator {
    associatedtype A
    
    func combine(_ x: A, _ y: A) -> A
}

protocol Transformer {
    associatedtype F
    
    func map<A, B>(_ fa: Kind<F, A>, _ f: @escaping (A) -> B) -> Kind<F, B>
}

// Instances

class IntCombinator: Combinator {
    typealias A = Int
    
    func combine(_ x: Int, _ y: Int) -> Int {
        return x + y
    }
}

extension Int {
    static var combinator: IntCombinator {
        return IntCombinator()
    }
}

class MaybeCombinator<V, CV>: Combinator where CV: Combinator, CV.A == V {
    typealias A = Maybe<V>
    
    let combinator: CV
    
    init(_ combinator: CV) {
        self.combinator = combinator
    }
    
    func combine(_ x: Maybe<V>, _ y: Maybe<V>) -> Maybe<V> {
        return x.fold(
            { y },
            { xx in y.fold(
                { x },
                { yy in Maybe.yes(combinator.combine(xx, yy))
            })
        })
    }
}

class MaybeTransformer: Transformer {
    typealias F = ForMaybe
    
    func map<A, B>(_ fa: Kind<ForMaybe, A>, _ f: @escaping (A) -> B) -> Kind<ForMaybe, B> {
        return fa.fix().fold({ Maybe<B>.no() },
                             { a in Maybe<B>.yes(f(a)) })
    }
}

extension Maybe {
    static func combinator<CV>(_ cv: CV) -> MaybeCombinator<A, CV> where CV: Combinator, CV.A == A {
        return MaybeCombinator(cv)
    }
    
    static func transformer() -> MaybeTransformer {
        return MaybeTransformer()
    }
}

// Program

struct Account {
    let id: String
    let balance: Int
}

class Program {
    let getBalanceBank1: Maybe<Account> = .yes(Account(id: "a1", balance: 100))
    let getBalanceBank2: Maybe<Int> = .yes(80)
    
    var balance: Maybe<Int> {
        let b1: Maybe<Int> = Maybe<Any>.transformer().map(getBalanceBank1, { acc in acc.balance }).fix()
        let b2: Maybe<Int> = getBalanceBank2
        
        return Maybe.combinator(Int.combinator).combine(b1, b2)
    }
    
    func run() {
        print(balance)
    }
}

Program().run()

File: kotlin/src/main/kotlin/com/fortyseven/fpfundamentals/Contents.kt
package com.fortyseven.fpfundamentals

import arrow.extension
import arrow.higherkind
import com.fortyseven.fpfundamentals.maybe.combinator.combinator

// Data types

@higherkind
sealed class Maybe<out A> : MaybeOf<A> {
  companion object {
  }

  data class Present<A>(val a: A) : Maybe<A>()
  object Absent : Maybe<Nothing>()

  inline fun <B> fold(ifAbsent: () -> B, ifPresent: (A) -> B): B = when (this) {
    is Absent -> ifAbsent()
    is Present<A> -> ifPresent(a)
  }
}

// Typeclasses

interface Combinator<A> {

  fun combine(x: A, y: A): A
}

// Instances

@extension
interface IntCombinator : Combinator<Int> {
  override fun combine(x: Int, y: Int): Int = x + y
}

fun Int.Companion.combinator(): Combinator<Int> =
  object : IntCombinator {}

@extension
interface MaybeCombinator<A> : Combinator<Maybe<A>> {

  fun CA(): Combinator<A>

  override fun combine(x: Maybe<A>, y: Maybe<A>): Maybe<A> {
    return x.fold(
      { y },
      { xx ->
        y.fold(
          { x },
          { yy -> Maybe.Present(CA().combine(xx, yy)) })
      })
  }
}

// Program

class Program {

  fun getBalanceBank1(): Maybe<Int> = Maybe.Present(100)

  fun getBalanceBank2(): Maybe<Int> = Maybe.Present(80)

  val balance: Maybe<Int>
    get() = Maybe.combinator(Int.combinator())
      .combine(getBalanceBank1(), getBalanceBank2())
}

fun main(args: Array<String>) {
  println(Program().balance)
}
